cpp/src/io/parquet/bloom_filter_reader.cu

```diff
@@ -0,0 +1,682 @@
+  cudf::device_span<void* const> buffer_ptrs;
+  cudf::device_span<size_t const> buffer_sizes;
+  host_span<Type const> parquet_types;
+  size_t num_row_groups;
+  size_t num_equality_columns;
+
+  template <typename T, bool timestamp_is_int96 = false>
+  std::unique_ptr<cudf::column> query_bloom_filter(cudf::size_type equality_col_idx,
+                                                   cudf::data_type dtype,
+                                                   ast::literal* const& literal,
+                                                   rmm::cuda_stream_view stream,
+                                                   rmm::device_async_resource_ref mr) const
+  {
+    using key_type    = T;
+    using hasher_type = cudf::hashing::detail::XXHash_64<key_type>;
+    using policy_type = cuco::arrow_filter_policy<key_type, hasher_type>;
+    using word_type   = typename policy_type::word_type;
+
+    // Check if the literal has the same type as the column
+    CUDF_EXPECTS(dtype.id() == literal->get_data_type().id(),
+                 "Mismatched data types between the column and the literal");
+
+    // Filter properties
+    auto constexpr word_size       = sizeof(word_type);
+    auto constexpr words_per_block = policy_type::words_per_block;
+
+    rmm::device_buffer results{num_row_groups, stream, mr};
+
+    // Query literal in bloom filters from each column chunk (row group).
+    thrust::for_each(
+      rmm::exec_policy_nosync(stream),
+      thrust::make_counting_iterator<size_t>(0),
+      thrust::make_counting_iterator(num_row_groups),
+      [buffer_ptrs          = buffer_ptrs.data(),
+       buffer_sizes         = buffer_sizes.data(),
+       d_scalar             = literal->get_value(),
+       col_idx              = equality_col_idx,
+       num_equality_columns = num_equality_columns,
+       results = reinterpret_cast<bool*>(results.data())] __device__(auto row_group_idx) {
+        // Filter bitset buffer index
+        auto const filter_idx = col_idx + (num_equality_columns * row_group_idx);
+
+        // If no bloom filter, then fill in `true` as membership cannot be determined
+        if (buffer_sizes[filter_idx] == 0) {
+          results[row_group_idx] = true;
+          return;
+        }
+
+        // Number of filter blocks
+        auto const num_filter_blocks = buffer_sizes[filter_idx] / (word_size * words_per_block);
+
+        // Create a bloom filter view.
+        cuco::bloom_filter_ref<key_type,
+                               cuco::extent<std::size_t>,
+                               cuco::thread_scope_thread,
+                               policy_type>
+          filter{reinterpret_cast<word_type*>(buffer_ptrs[filter_idx]),
+                 num_filter_blocks,
+                 {},  // Thread scope as the same literal is being searched across different bitsets
+                      // per thread
+                 {}};  // Arrow policy with cudf::XXHash_64 seeded with 0 for Arrow compatibility
```

Comment:
`cudf::XXHash_64` isn't a public type. Does this need to say `cudf::hashing::detail::XXHash_64`?

Reply:
```suggestion
                 {}};  // Arrow policy with cudf::hashing::detail::XXHash_64 seeded with 0 for Arrow compatibility
```